Cython/Utility/Coroutine.c

```diff
@@ -993,65 +1092,76 @@ static int __Pyx_Coroutine_CloseIter(__pyx_CoroutineObject *gen, PyObject *yf) {
         } else {
             retval = __Pyx_PyObject_CallNoArg(meth);
             Py_DECREF(meth);
-            if (unlikely(!retval))
-                err = -1;
+            if (unlikely(!retval)) {
+                result = PYGEN_ERROR;
+            }
         }
         gen->is_running = 0;
     }
     Py_XDECREF(retval);
-    return err;
+    return result == PYGEN_ERROR ? -1 : 0;
 }
 
 static PyObject *__Pyx_Generator_Next(PyObject *self) {
+    __Pyx_PySendResult result;
+    PyObject *retval = NULL;
     __pyx_CoroutineObject *gen = (__pyx_CoroutineObject*) self;
-    PyObject *yf = gen->yieldfrom;
-    if (unlikely(gen->is_running))
+    if (unlikely(gen->is_running)) {
         return __Pyx_Coroutine_AlreadyRunningError(gen);
-    if (yf) {
+    }
+    if (gen->yieldfrom_am_send) {
+        result = __Pyx_Coroutine_SendToDelegate(gen, gen->yieldfrom_am_send, Py_None, &retval);
+    } else if (gen->yieldfrom) {
+        PyObject *yf = gen->yieldfrom;
         PyObject *ret;
-        // FIXME: does this really need an INCREF() ?
-        //Py_INCREF(yf);
         // YieldFrom code ensures that yf is an iterator
         gen->is_running = 1;
         #ifdef __Pyx_Generator_USED
         if (__Pyx_Generator_CheckExact(yf)) {
             ret = __Pyx_Generator_Next(yf);
         } else
         #endif
-        #if CYTHON_COMPILING_IN_CPYTHON
+        #if CYTHON_COMPILING_IN_CPYTHON && PY_VERSION_HEX < 0x030A00A3
+        // _PyGen_Send() is not needed in 3.10+ due to "am_send"
         if (PyGen_CheckExact(yf)) {
             ret = __Pyx_PyGen_Send((PyGenObject*)yf, NULL);
         } else
-        #endif
-        #ifdef __Pyx_Coroutine_USED
-        if (__Pyx_Coroutine_Check(yf)) {
-            ret = __Pyx_Coroutine_Send(yf, Py_None);
-        } else
         #endif
             ret = __Pyx_PyObject_GetIterNextFunc(yf)(yf);
         gen->is_running = 0;
-        //Py_DECREF(yf);
         if (likely(ret)) {
             return ret;
         }
-        return __Pyx_Coroutine_FinishDelegation(gen);
+        result = __Pyx_Coroutine_FinishDelegation(gen, &retval);
+    } else {
+        result = __Pyx_Coroutine_SendEx(gen, Py_None, &retval, 0);
     }
-    return __Pyx_Coroutine_SendEx(gen, Py_None, 0);
+
+    return __Pyx_Coroutine_MethodReturnFromResult(self, result, retval);
 }
 
 static PyObject *__Pyx_Coroutine_Close_Method(PyObject *self, PyObject *arg) {
+    PyObject *retval = NULL;
+    __Pyx_PySendResult result;
     CYTHON_UNUSED_VAR(arg);
-    return __Pyx_Coroutine_Close(self);
+    result = __Pyx_Coroutine_Close(self, &retval);
+    if (unlikely(result == PYGEN_ERROR))
+        return NULL;
+    Py_XDECREF(retval);
+    Py_RETURN_NONE;
 }
 
-static PyObject *__Pyx_Coroutine_Close(PyObject *self) {
+static __Pyx_PySendResult
+__Pyx_Coroutine_Close(PyObject *self, PyObject **retval) {
```

Comment:
It's possible I've missed something, but I don't think any of the callers use `retval` so it feels like an complication to return it.